File: codeforces/practice/main_sequence.cpp
//https://codeforces.com/contest/286/problem/C
#include<bits/stdc++.h>
using namespace std;
const int N = 1e6+10;
map<int,set<int>> m;
int arr[N], fix[N];

int main(){
    int a,n,d;
    scanf("%d",&n);
    for(int i = 1; i <= n; ++i){
        scanf("%d",arr+i);
    }
    reverse(arr+1,arr+n+1);
    scanf("%d",&d);
    while(d--){
        scanf("%d",&a);
        arr[n-a+1] = -arr[n-a+1];
    }
    stack<int> s;
    for(int i = 1; i <= n; ++i){
        if(arr[i] > 0 && s.size() && arr[s.top()] == -arr[i])s.pop();
        else{
            s.push(i);
            arr[i] = -abs(arr[i]);
        }
    }
    if(s.size())return printf("NO"),0;
    printf("YES\n");
    reverse(arr+1,arr+n+1);
    for(int i = 1; i <= n; ++i)printf("%d ",arr[i]);
    return 0;
}

File: codeforces/practice/main_sequence.java
import java.util.*;
import java.io.*;
public class main_sequence{
    public static void main(String[] args)  throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] arr = new int[n];
        for(int i = 0; i < n; ++i)arr[i] = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int neg = Integer.parseInt(st.nextToken());
        if(neg*2 > n){
            System.out.println("NO");
            return;
        }
        for(int i = 0; i < n/2; ++i){
            int e = arr[i];
            arr[i] = arr[n-1-i];
            arr[n-1-i] = e;
        }
        for(int i = 0; i < neg; ++i){
            int ind = Integer.parseInt(st.nextToken());
            ind--;
            arr[n-1-ind] = -arr[n-1-ind];
        }
        Stack<Integer> s = new Stack<Integer> ();
        for(int i = 0; i < n; ++i){
            if(arr[i] > 0 && !s.empty() && arr[s.peek()] == -arr[i])s.pop();
            else{
                arr[i] = -Math.abs(arr[i]);
                s.push(i);
            }
        }
        if(!s.empty()){
            System.out.println("NO");
            return;
        }
        System.out.println("YES");
        StringBuilder res = new StringBuilder("");
        for(int i = n-1; i >= 0; --i)res.append(Integer.toString(arr[i]) + " ");
        System.out.println(res.toString());
    }
}